packages/intrepid2/src/Discretization/Basis/Intrepid2_IntegratedLegendreBasis_HGRAD_TET.hpp

```diff
@@ -0,0 +1,850 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov),
+//                    Mauro Perego  (mperego@sandia.gov), or
+//                    Nate Roberts  (nvrober@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+/** \file   Intrepid2_IntegratedLegendreBasis_HGRAD_TET.hpp
+    \brief  H(grad) basis on the tetrahedon based on integrated Legendre polynomials.
+    \author Created by N.V. Roberts.
+ */
+
+#ifndef Intrepid2_IntegratedLegendreBasis_HGRAD_TET_h
+#define Intrepid2_IntegratedLegendreBasis_HGRAD_TET_h
+
+#include <Kokkos_View.hpp>
+#include <Kokkos_DynRankView.hpp>
+
+#include <Intrepid2_config.h>
+
+#include "Intrepid2_Polynomials.hpp"
+#include "Intrepid2_Utils.hpp"
+
+namespace Intrepid2
+{
+  /** \class  Intrepid2::Hierarchical_HGRAD_TET_Functor
+      \brief  Functor for computing values for the IntegratedLegendreBasis_HGRAD_TET class.
+   
+   This functor is not intended for use outside of IntegratedLegendreBasis_HGRAD_TET.
+  */
+  template<class ExecutionSpace, class OutputScalar, class PointScalar,
+           class OutputFieldType, class InputPointsType>
+  struct Hierarchical_HGRAD_TET_Functor
+  {
+    using ScratchSpace        = Kokkos::DefaultExecutionSpace::scratch_memory_space;
+    using OutputScratchView   = Kokkos::View<OutputScalar*,ScratchSpace,Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using OutputScratchView2D = Kokkos::View<OutputScalar**,ScratchSpace,Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using PointScratchView    = Kokkos::View<PointScalar*, ScratchSpace,Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    
+    using TeamPolicy = Kokkos::TeamPolicy<>;
+    using TeamMember = TeamPolicy::member_type;
+    
+    EOperator opType_;
+    
+    OutputFieldType  output_;      // F,P
+    InputPointsType  inputPoints_; // P,D
+    
+    int polyOrder_;
+    bool defineVertexFunctions_;
+    int numFields_, numPoints_;
+    
+    size_t fad_size_output_;
+    
+    static const int numVertices = 4;
+    static const int numEdges    = 6;
+    // the following ordering of the edges matches that used by ESEAS
+    const int edge_start_[numEdges] = {0,1,0,0,1,2}; // edge i is from edge_start_[i] to edge_end_[i]
+    const int edge_end_[numEdges]   = {1,2,2,3,3,3}; // edge i is from edge_start_[i] to edge_end_[i]
+    
+    static const int numFaces    = 4;
+    const int face_vertex_0[numFaces] = {0,0,1,0}; // faces are abc where 0 ≤ a < b < c ≤ 3
+    const int face_vertex_1[numFaces] = {1,1,2,2};
+    const int face_vertex_2[numFaces] = {2,3,3,3};
+    
```

Comment:
Ah, yes, I neglected to consider when setting up the tag to ordinal data.

Reading the code in `Shards_BasicTopologies.hpp` around line 687 (unfortunately, no nice ASCII art for tetrahedron!), it looks like the edge ordering happens to agree between ESEAS and Shards, but the faces are different.  It looks like the shards face ordering is:
```
013
123
032
021
```
and the ESEAS ordering is:
```
012
013
123
023
```
I'll work both on fixing the tag to ordinal map, and refining the basis equivalence test to optionally check that dofs on one face in basis1 are represented in terms of dofs on that face in basis2.  (Optionally because the "DG" bases are all interior functions; this only makes sense when comparing two CG bases.)

Reply:
(Thanks for catching this!)

Reply:
OK, I think I've now resolved this (including a test).